File: convert_to_markdown.py
#!/usr/bin/env python3
"""
Script to convert .docx and .rtf files to Markdown while preserving formatting.
"""

import os
import sys
from pathlib import Path
import re

try:
    from docx import Document
except ImportError:
    print("Installing python-docx...")
    os.system(f"{sys.executable} -m pip install python-docx")
    from docx import Document

try:
    from striprtf.striprtf import rtf_to_text
except ImportError:
    print("Installing striprtf...")
    os.system(f"{sys.executable} -m pip install striprtf")
    from striprtf.striprtf import rtf_to_text


def convert_docx_to_markdown(docx_path):
    """Convert a .docx file to Markdown preserving formatting."""
    try:
        doc = Document(docx_path)
        markdown_lines = []
        
        for paragraph in doc.paragraphs:
            if not paragraph.text.strip():
                markdown_lines.append("")
                continue
            
            # Process runs to preserve formatting
            text_parts = []
            for run in paragraph.runs:
                text = run.text
                if not text:
                    continue

                # Apply formatting using pure Markdown (no HTML)
                if run.bold:
                    text = f"**{text}**"
                if run.italic:
                    text = f"*{text}*"
                if run.underline:
                    text = f"__{text}__"

                text_parts.append(text)
            
            # If there are no formatted runs, fall back to the paragraph text
            if not text_parts:
                para_text = paragraph.text
            else:
                para_text = "".join(text_parts)
            
            # Inspect paragraph style
            style = paragraph.style.name.lower()
            
            # Headings
            if 'heading' in style or 'title' in style:
                level = 1
                if 'heading 1' in style or 'title' in style:
                    level = 1
                elif 'heading 2' in style:
                    level = 2
                elif 'heading 3' in style:
                    level = 3
                elif 'heading 4' in style:
                    level = 4
                elif 'heading 5' in style:
                    level = 5
                elif 'heading 6' in style:
                    level = 6
                
                markdown_lines.append(f"{'#' * level} {para_text}")
            else:
                markdown_lines.append(para_text)
        
        # Process tables
        for table in doc.tables:
            markdown_lines.append("")
            # Header
            header_row = table.rows[0]
            header_cells = [cell.text.strip() for cell in header_row.cells]
            markdown_lines.append("| " + " | ".join(header_cells) + " |")
            markdown_lines.append("| " + " | ".join(["---"] * len(header_cells)) + " |")
            
            # Data rows
            for row in table.rows[1:]:
                cells = [cell.text.strip() for cell in row.cells]
                markdown_lines.append("| " + " | ".join(cells) + " |")
            markdown_lines.append("")
        
        return "\n".join(markdown_lines)
    
    except Exception as e:
        return f"# Error converting {docx_path}\n\nError: {str(e)}"


def convert_rtf_to_markdown(rtf_path):
    """Convert an .rtf file to Markdown preserving basic formatting."""
    try:
        with open(rtf_path, 'rb') as f:
            rtf_content = f.read()
        
        # Try different encodings
        encodings = ['latin-1', 'cp1252', 'iso-8859-1', 'utf-8']
        rtf_text = None
        
        for encoding in encodings:
            try:
                candidate = rtf_content.decode(encoding, errors='ignore')
                parsed = rtf_to_text(candidate)
                if parsed.strip():
                    rtf_text = candidate
                    break
            except Exception:
                continue
        
        if rtf_text is None:
            rtf_text = rtf_content.decode('latin-1', errors='ignore')
        
        # Use striprtf to extract clean text
        try:
            plain_text = rtf_to_text(rtf_text)
        except Exception as e:
            # Fallback: basic manual extraction removing RTF commands
            plain_text = rtf_text
            # Remove empty RTF groups first
            while '{' in plain_text and '}' in plain_text:
                plain_text = re.sub(r'\{[^{}]*\}', '', plain_text)
            # Remove RTF commands
            plain_text = re.sub(r'\\[a-z]+\d*\s*', ' ', plain_text)
            plain_text = re.sub(r'\\[{}]', '', plain_text)
            # Remove RTF special characters
            plain_text = re.sub(r'\\\'[0-9a-f]{2}', '', plain_text)
            # Remove stray numbers that belong to RTF commands
            plain_text = re.sub(r'\s+\d+\s+', ' ', plain_text)
        
        # Clean extracted text
        # Remove lines that are only font names or commands
        lines = plain_text.split('\n')
        cleaned_lines = []
        
        for line in lines:
            line = line.strip()
            
            # Strip control characters
            line = re.sub(r'[\x00-\x1f\x7f-\x9f]', '', line)
            line = line.strip()
            
            # Drop lines that are only artifacts
            if not line:
                cleaned_lines.append("")
                continue
            
            # Remove lines that are only font names or RTF commands
            if (line.lower() in ['times new roman', 'arial', 'calibri', 'helvetica', 
                                 'trebuchet ms', 'cambria', 'times'] or
                re.match(r'^[a-z\s]+\}?$', line.lower()) or
                re.match(r'^[\d\s\-]+$', line) or
                line.count('}') > line.count(' ') or
                (len(line) < 3 and not line.isalnum())):
                continue
            
            # Remove RTF conversion artifacts
            line = re.sub(r'\s+', ' ', line)  # Multiple spaces
            line = re.sub(r'^[a-z]+\s+[a-z]+\s+', '', line)  # Remove stray words at start
            line = re.sub(r'\s*\}\s*$', '', line)  # Remove braces at end
            line = re.sub(r'^\s*\{\s*', '', line)  # Remove braces at start
            
            # Remove stray numbers at start/end
            line = re.sub(r'^\s*[\d\-]+\s+', '', line)
            line = re.sub(r'\s+[\d\-]+\s*$', '', line)
            
            if line.strip():
                cleaned_lines.append(line.strip())
        
        # Process cleaned lines and apply formatting
        markdown_lines = []
        
        for line in cleaned_lines:
            if not line:
                markdown_lines.append("")
                continue
            
            # Detect primary headings (uppercase, no trailing period, contains keywords)
            if (line.isupper() and 15 < len(line) < 120 and 
                not line.endswith('.') and 
                ('TOMOGRAFIA' in line or 'ANGIO' in line or 'COMPUTADORIZADA' in line)):
                markdown_lines.append(f"## {line}")
            # Detect important sections
            elif any(keyword in line.upper() for keyword in 
                    ['INDICAÇÃO CLÍNICA', 'TÉCNICA DO EXAME', 'ASPECTOS OBSERVADOS', 'IMPRESSÃO']):
                # Uppercase text is a heading
                if line.isupper() and len(line) > 10:
                    markdown_lines.append(f"## {line}")
                else:
                    # If it starts with a keyword, make it bold
                    for keyword in ['INDICAÇÃO', 'TÉCNICA', 'ASPECTOS', 'IMPRESSÃO']:
                        if line.upper().startswith(keyword):
                            markdown_lines.append(f"**{line}**")
                            break
                    else:
                        markdown_lines.append(line)
            # Detect italic text (usually footnotes)
            elif ('probabilidade' in line.lower() or 
                  'médico' in line.lower() or 
                  'diagnóstica' in line.lower()):
                markdown_lines.append(f"*{line}*")
            else:
                markdown_lines.append(line)
        
        # Remove excessive empty lines
        result = []
        prev_empty = False
        for line in markdown_lines:
            if not line.strip():
                if not prev_empty:
                    result.append("")
                prev_empty = True
            else:
                result.append(line)
                prev_empty = False
        
        return "\n".join(result)
    
    except Exception as e:
        import traceback
        return f"# Error converting {rtf_path}\n\nError: {str(e)}\n\nTraceback: {traceback.format_exc()}"


def main():
    """Main function that processes all files inside the Reports folder."""
    reports_dir = Path(__file__).parent / "Templates_docx"
    
    if not reports_dir.exists():
        print(f"Error: Folder {reports_dir} not found!")
        return
    
    # Create folder for Markdown output
    markdown_dir = reports_dir.parent / "Templates_markdown"
    markdown_dir.mkdir(exist_ok=True)
    
    # Process .docx files
    docx_files = list(reports_dir.glob("*.docx"))
    print(f"Found {len(docx_files)} .docx files")
    
    for docx_file in docx_files:
        print(f"Converting {docx_file.name}...")
        markdown_content = convert_docx_to_markdown(docx_file)
        output_file = markdown_dir / f"{docx_file.stem}.md"
        with open(output_file, 'w', encoding='utf-8') as f:
            f.write(markdown_content)
        print(f"  ✓ Saved to {output_file.name}")
    
    # Process .rtf files
    rtf_files = list(reports_dir.glob("*.rtf"))
    print(f"\nFound {len(rtf_files)} .rtf files")
    
    for rtf_file in rtf_files:
        print(f"Converting {rtf_file.name}...")
        markdown_content = convert_rtf_to_markdown(rtf_file)
        output_file = markdown_dir / f"{rtf_file.stem}.md"
        with open(output_file, 'w', encoding='utf-8') as f:
            f.write(markdown_content)
        print(f"  ✓ Saved to {output_file.name}")
    
    print(f"\n✓ Conversion finished! Files saved to {markdown_dir}")


if __name__ == "__main__":
    main()

File: rust_converters/src/bin/convert_to_markdown.rs
use anyhow::Result;
use docx_rust::document::{BodyContent, Paragraph, ParagraphContent};
use docx_rust::formatting::{Bold, Italics, Underline, UnderlineStyle};
use docx_rust::DocxFile;
use regex::Regex;
use std::fs;
use std::path::{Path, PathBuf};

fn convert_docx_to_markdown(docx_path: &Path) -> Result<String> {
    let file = DocxFile::from_file(docx_path)?;
    let docx = file.parse()?;

    let mut markdown_lines: Vec<String> = Vec::new();

    // Walk the document body in order.
    // Only handle paragraphs; tables/SDT/etc. are ignored.
    let body = &docx.document.body;
    for item in &body.content {
        match item {
            BodyContent::Paragraph(p) => {
                markdown_lines.push(paragraph_to_markdown(p));
            }
            // Se quiser mapear tabelas em Markdown, implementar aqui
            BodyContent::Table(_t) => {
                // Ignored for simplicity — the Python version outputs table Markdown.
            }
            _ => {}
        }
    }

    Ok(markdown_lines.join("\n"))
}

fn bold_is_on(flag: &Option<Bold>) -> bool {
    flag.as_ref()
        .map(|b| b.value.unwrap_or(true))
        .unwrap_or(false)
}

fn italics_is_on(flag: &Option<Italics>) -> bool {
    flag.as_ref()
        .map(|i| i.value.unwrap_or(true))
        .unwrap_or(false)
}

fn underline_is_on(flag: &Option<Underline>) -> bool {
    flag.as_ref()
        .map(|u| match u.val.as_ref() {
            Some(UnderlineStyle::None) => false,
            Some(_) => true,
            None => true, // element is present, default style is "single"
        })
        .unwrap_or(false)
}

fn paragraph_to_markdown(p: &Paragraph) -> String {
    let plain = p.text();
    if plain.trim().is_empty() {
        return String::new();
    }

    // Process runs to preserve bold/italic/underline.
    let mut text_parts: Vec<String> = Vec::new();
    for pc in &p.content {
        if let ParagraphContent::Run(run) = pc {
            let mut text = run.text();
            if text.is_empty() {
                continue;
            }

            if let Some(prop) = &run.property {
                if bold_is_on(&prop.bold) {
                    text = format!("**{}**", text);
                }
                if italics_is_on(&prop.italics) {
                    text = format!("*{}*", text);
                }
                if underline_is_on(&prop.underline) {
                    text = format!("__{}__", text);
                }
            }

            text_parts.push(text);
        }
    }

    if text_parts.is_empty() {
        // Fall back to the aggregate paragraph text
        plain
    } else {
        text_parts.join("")
    }
}

fn convert_rtf_to_markdown(rtf_path: &Path) -> Result<String> {
    let bytes = fs::read(rtf_path)?;
    // Python tries multiple encodings; here we take a simpler step.
    let mut rtf_text = String::from_utf8_lossy(&bytes).to_string();

    // Remove simple RTF groups { ... } (no deep nesting)
    let re_group = Regex::new(r"\{[^{}]*\}")?;
    while rtf_text.contains('{') && rtf_text.contains('}') {
        let new = re_group.replace_all(&rtf_text, "");
        let new_owned = new.into_owned();
        if new_owned == rtf_text {
            break;
        }
        rtf_text = new_owned;
    }

    // Remove simple RTF commands \wordN?
    let re_cmd = Regex::new(r"\\[a-zA-Z]+\d*\s*")?;
    rtf_text = re_cmd.replace_all(&rtf_text, " ").into_owned();

    // Remove brace escapes
    let re_brace_cmd = Regex::new(r"\\[{}]")?;
    rtf_text = re_brace_cmd.replace_all(&rtf_text, "").into_owned();

    // Remove hex escapes \\'hh
    let re_hex = Regex::new(r"\\'[0-9a-fA-F]{2}")?;
    rtf_text = re_hex.replace_all(&rtf_text, "").into_owned();

    // Remove loose numbers from commands
    let re_nums = Regex::new(r"\s+\d+\s+")?;
    rtf_text = re_nums.replace_all(&rtf_text, " ").into_owned();

    let mut cleaned_lines: Vec<String> = Vec::new();

    for raw_line in rtf_text.lines() {
        let mut line = raw_line.trim().to_string();

        // Remove control characters
        line = line
            .chars()
            .filter(|c| !(*c as u32 <= 0x1F || (0x7F..=0x9F).contains(&(*c as u32))))
            .collect::<String>()
            .trim()
            .to_string();

        if line.is_empty() {
            cleaned_lines.push(String::new());
            continue;
        }

        let lower = line.to_lowercase();

        // Common fonts / artifacts
        let is_font_name = matches!(
            lower.as_str(),
            "times new roman"
                | "arial"
                | "calibri"
                | "helvetica"
                | "trebuchet ms"
                | "cambria"
                | "times"
        );

        let re_only_words = Regex::new(r"^[a-z\s]+\}?$").unwrap();
        let re_only_nums = Regex::new(r"^[\d\s\-]+$").unwrap();

        if is_font_name
            || re_only_words.is_match(&lower)
            || re_only_nums.is_match(&line)
            || line.chars().filter(|c| *c == '}').count()
                > line.chars().filter(|c| *c == ' ').count()
            || (line.len() < 3 && !line.chars().all(|c| c.is_alphanumeric()))
        {
            continue;
        }

        // Additional cleanup
        let re_spaces = Regex::new(r"\s+").unwrap();
        line = re_spaces.replace_all(&line, " ").into_owned();

        let re_stray_words_start = Regex::new(r"^[a-zA-Z]+\s+[a-zA-Z]+\s+").unwrap();
        line = re_stray_words_start.replace_all(&line, "").into_owned();

        let re_brace_end = Regex::new(r"\s*\}\s*$").unwrap();
        line = re_brace_end.replace_all(&line, "").into_owned();

        let re_brace_start = Regex::new(r"^\s*\{\s*").unwrap();
        line = re_brace_start.replace_all(&line, "").into_owned();

        let re_nums_start = Regex::new(r"^\s*[\d\-]+\s+").unwrap();
        line = re_nums_start.replace_all(&line, "").into_owned();

        let re_nums_end = Regex::new(r"\s+[\d\-]+\s*$").unwrap();
        line = re_nums_end.replace_all(&line, "").into_owned();

        if !line.trim().is_empty() {
            cleaned_lines.push(line.trim().to_string());
        }
    }

    // Markdown formatting heuristics
    let mut markdown_lines: Vec<String> = Vec::new();

    for line in cleaned_lines {
        if line.is_empty() {
            markdown_lines.push(String::new());
            continue;
        }

        let upper = line.to_uppercase();
        let lower = line.to_lowercase();

        // Main heading (all caps, no trailing period, contains keywords)
        if upper == line
            && line.len() > 15
            && line.len() < 120
            && !line.ends_with('.')
            && (upper.contains("TOMOGRAFIA")
                || upper.contains("ANGIO")
                || upper.contains("COMPUTADORIZADA"))
        {
            markdown_lines.push(format!("## {}", line));
            continue;
        }

        // Important sections
        let keywords = [
            "indicação clínica",
            "técnica do exame",
            "aspectos observados",
            "impressão",
        ];

        if keywords.iter().any(|k| lower.contains(k)) {
            if upper == line && line.len() > 10 {
                markdown_lines.push(format!("## {}", line));
                continue;
            } else {
                let start_keywords = ["indicação", "técnica", "aspectos", "impressão"];
                let mut handled = false;
                for k in &start_keywords {
                    if lower.starts_with(k) {
                        markdown_lines.push(format!("**{}**", line));
                        handled = true;
                        break;
                    }
                }
                if !handled {
                    markdown_lines.push(line);
                }
                continue;
            }
        }

        // Footnote-like notes → italic
        if lower.contains("probabilidade")
            || lower.contains("médico")
            || lower.contains("diagnóstica")
        {
            markdown_lines.push(format!("*{}*", line));
        } else {
            markdown_lines.push(line);
        }
    }

    // Remove duplicate empty lines
    let mut result: Vec<String> = Vec::new();
    let mut prev_empty = false;
    for line in markdown_lines {
        if line.trim().is_empty() {
            if !prev_empty {
                result.push(String::new());
                prev_empty = true;
            }
        } else {
            result.push(line);
            prev_empty = false;
        }
    }

    Ok(result.join("\n"))
}

fn main() -> Result<()> {
    let reports_dir = PathBuf::from("Templates_docx");
    if !reports_dir.exists() {
        eprintln!("Error: Folder {} not found!", reports_dir.display());
        return Ok(());
    }

    let markdown_dir = reports_dir
        .parent()
        .unwrap_or_else(|| Path::new("."))
        .join("Templates_markdown");
    fs::create_dir_all(&markdown_dir)?;

    // Process .docx
    let docx_files: Vec<PathBuf> = fs::read_dir(&reports_dir)?
        .filter_map(|e| e.ok())
        .map(|e| e.path())
        .filter(|p| p.extension().and_then(|s| s.to_str()) == Some("docx"))
        .collect();

    println!("Found {} .docx files", docx_files.len());
    for docx_file in &docx_files {
        println!(
            "Converting {}...",
            docx_file.file_name().unwrap().to_string_lossy()
        );
        let markdown_content = convert_docx_to_markdown(docx_file)?;
        let output_file = markdown_dir.join(
            docx_file
                .file_stem()
                .unwrap()
                .to_string_lossy()
                .to_string()
                + ".md",
        );
        fs::write(&output_file, markdown_content)?;
        println!("  ✓ Saved to {}", output_file.file_name().unwrap().to_string_lossy());
    }

    // Process .rtf
    let rtf_files: Vec<PathBuf> = fs::read_dir(&reports_dir)?
        .filter_map(|e| e.ok())
        .map(|e| e.path())
        .filter(|p| p.extension().and_then(|s| s.to_str()) == Some("rtf"))
        .collect();

    println!("\nFound {} .rtf files", rtf_files.len());
    for rtf_file in &rtf_files {
        println!(
            "Converting {}...",
            rtf_file.file_name().unwrap().to_string_lossy()
        );
        let markdown_content = convert_rtf_to_markdown(rtf_file)?;
        let output_file = markdown_dir.join(
            rtf_file
                .file_stem()
                .unwrap()
                .to_string_lossy()
                .to_string()
                + ".md",
        );
        fs::write(&output_file, markdown_content)?;
        println!("  ✓ Saved to {}", output_file.file_name().unwrap().to_string_lossy());
    }

    println!("\n✓ Conversion finished! Files saved to {}", markdown_dir.display());
    Ok(())
}
